### ISSUE TITLE
allow a batch job to be restarted

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#4912]
Problem: Flux batch jobs are not restartable from a dump

This is pretty close to being a possibility given that the system instance is restartable

Some obstacles are
- restarting on a new resource set
- restarting the batch script (or guidelines/tools for making a batch script restartable)
- coordinating a clean stop (avoid killing running jobs)

**Comment by grondo** [comment]
Some early thoughts:

> restarting on a new resource set

One proposal is to restore only the `job.*` kvs dir on restart. This would work now even if the job was run on a different number of nodes. Essentially this solves the "preserving the queue" use case.

> restarting the batch script (or guidelines/tools for making a batch script restartable)

For an initial deliverable, what if batch job restarts do not re-run the batch script, but instead use a standard script or utility that simply waits for all jobs to complete?  So an initial batch script submits workload and restarts just continue running that workload until it is complete. I think this is the only real use case we've heard for this feature. For users that are running a single parallel job per batch job, that job itself is usually restartable from its own dump files, so there isn't really a need for a restartable Flux instance AFAIK.

> coordinating a clean stop (avoid killing running jobs)

Can you expand on that item? The use case for restarting a batch job I had in mind was to continue workloads that were terminated due to timeout or node failure. I'm probably missing a use case where a clean stop is necessary (if the user wants a continuously running batch script, why would they elect to stop it?)

I guess one missing obstacle is automatically restarting jobs that failed to run to completion in a previous incarnation of the batch job.

**Comment by garlick** [comment]
> One proposal is to restore only the job.* kvs dir on restart. 

We also would want to preserve `admin.startlog`  and `checkpoint.job-manager`.  I think the main problem there is that  `resource.eventlog` records drained nodes that might not be part of the instance anymore (worse they are currently recorded by rank) so perhaps we could just omit that for a start.

> For an initial deliverable, what if batch job restarts do not re-run the batch script, but instead use a standard script or utility that simply waits for all jobs to complete

:+1: although in that case we have to stipulate that the user's batch script can _only_ submit work and block waiting for it to finish.  It can't do any cleanup or post processing after the work finishes, and it can't block anywhere in the middle to start a new phase or steer computation or whatever.

> Can you expand on that item? The use case for restarting a batch job I had in mind was to continue workloads that were terminated due to timeout or node failure. I'm probably missing a use case where a clean stop is necessary (if the user wants a continuously running batch script, why would they elect to stop it?)

Well, I was thinking that someone might elect to run a long parameter study as a restartable batch job submitted for the max duration multiple times and that it might be advantageous to not kill jobs if possible.  But I had forgotten that fluxion will already make a schedule that fits in the allotted time as long as each sub-job's worst case runtime can be estimated at submission.  So maybe a non-issue.

Another use case for serial allocations might be if  a batch job contains a workload of big and small jobs that can complete in any order.  The the batch job might be submitted to run first on a small allocation, then on progressively bigger allocations, like a cheesy version of "grow".

How would failed jobs be automatically resubmitted/restarted in the restarted batch instance?

**Comment by grondo** [comment]
> We also would want to preserve admin.startlog and checkpoint.job-manager

Ok, I can see how `admin.startlog` is interesting, though it doesn't seem necessary, but isn't the job-manager checkpoint only written out on a clean shutdown? Won't we have to be able to restart without it anyway in case a batch job crashes?

> although in that case we have to stipulate that the user's batch script can only submit work and block waiting for it to finish

Yeah, this is why I considered just an initial deliverable, o/w the batch script would have to be broken up into parts.

> Another use case for serial allocations might be if a batch job contains a workload of big and small jobs that can complete in any order. The the batch job might be submitted to run first on a small allocation, then on progressively bigger allocations, like a cheesy version of "grow".

Yeah, that is the kind of use case I had in mind as well. I'm not sure how this differs from submitting any number of jobs that can't run to completion in an existing batch job, whether that is due to the max runtime of the batch job or its size. One other issue this does bring up is that Fluxion will reject jobs that are unsatisfiable in the current instance, so we'd need some way to prevent that as well for this particular use case.

> How would failed jobs be automatically resubmitted/restarted in the restarted batch instance?

Yeah, that is an open question. I was proposing you add that to the checklist for this issue above.

It would be trivial to write a script or command that resubmits jobs that failed with a certain class of exceptions, but perhaps we'd want to think more generally about how jobs could be marked as "restartable" or "requeue-able" (e.g. could the jobid and kvs dir be reused? etc)

**Comment by grondo** [comment]
Would it make sense to include the jobid in the startlog for batch jobs? That might make the `admin.startlog` a bit more interesting since you could tie each "start" back to a jobid in the parent instance.

**Comment by garlick** [comment]
> isn't the job-manager checkpoint only written out on a clean shutdown? Won't we have to be able to restart without it anyway in case a batch job crashes?

A dump is only written out during a clean shutdown...

> Yeah, that is an open question. I was proposing you add that to the checklist for this issue above.

Apologies, I has missed your comment on the first read.

**Comment by grondo** [comment]
> A dump is only written out during a clean shutdown...

That should have been obvious, sorry!

This is only really compelling then in the case where jobs can be cleanly shutdown. We may want to think about how to handle batch jobs that are canceled by the system or lost due to a node crash, which would be a much more compelling use case. Perhaps a way to recover from an existing content backing store, or a method to incrementally update a defensive dumpfiles periodically during a job.

**Comment by garlick** [comment]
Batch jobs canceled by the system but otherwise healthy _should_ get a clean shutdown.

**Comment by grondo** [comment]
Good point. We might want to test that. Currently a job that is canceled or times out only has a 5 second grace time until SIGKILL after SIGTERM or SIGALRM.

**Comment by grondo** [comment]
Oh, yeah, that works in simple tests, we may want to reexamine the gracetime given to jobs just in case.

So, is handling a rank 0 node or broker failure just future work, or not possible using this scheme?

**Comment by garlick** [comment]
> So, is handling a rank 0 node or broker failure just future work, or not possible using this scheme?

I think future work involving either periodic defensive dumps or arranging for the sqlite db to be preserved.
